### ISSUE TITLE
Seamlessly handle callback with extra attribute as regular function/methods

### ISSUE THREAD

**Issue by nicolasdespres** [microsoft/pyright#10695]
Hello,

Thanks for creating such a great tool! It is amazing.

I wrote a decorator that just adds an attribute to the decorated function/method, and I am having a hard time making it properly handled by pyright.

```python
import functools
from typing import Callable, Generic, TypeVar, ParamSpec, reveal_type

_PWrapped = ParamSpec("_PWrapped")
_RWrapped = TypeVar("_RWrapped")

# Inspired from _Wrapper in functools.pyi
class CallableWithAttr(Generic[_PWrapped, _RWrapped]):
    attr: Callable[_PWrapped, _RWrapped]
    def __call__(self, *args: _PWrapped.args, **kwargs: _PWrapped.kwargs) -> _RWrapped: ...
    # as with ``Callable``, we'll assume that these attributes exist
    __name__: str
    __qualname__: str

def add_attr[**P, R](fn: Callable[P, R]) -> CallableWithAttr[P, R]:
    @functools.wraps(fn)
    def wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
        return fn(*args, **kwargs)
    wrapper.attr = fn
    return wrapper

@add_attr
def f(x: int, y: int) -> int:
    """A simple function that adds two integers."""
    return x + y

# Properly report the incorrect first argument type, and says that f is a function.
# But unfortunately it does not show the docstring of the function, nor its signature using the usual presentation.
# It only shows its type (CallableWithAttr), which is correct but not helpful.
f('a', 4)
reveal_type(f.attr)   # Properly recognizes the attribute.

class A:
    @add_attr
    def f(self, x: int, y: int) -> int:
        """A simple function that adds two integers."""
        return x + y

a = A()
# Properly recognizes it as a method, but expects the self argument to be explicitly passed,
# and does not show the docstring of the method using the usual presentation. I think it is a bug, unless my
# decorator is faulty.
print(a.f('a', 4))
```

What I would expect is two-fold:
1. The self argument should be properly handled when decorating a method (is it a bug?)
2. I would expect to have the same usual tooltip describing the decorated function when I call it: its full signature and its docstring. Its true type could be added as a simple note at then end of the tooltip.

I tried the solution offered in this [thread](https://stackoverflow.com/questions/62658540/how-to-combine-a-custom-protocol-with-the-callable-protocol).
I am using Python 3.12.

Thanks for reading!

**Comment by erictraut** [comment]
There are two independent parts here:
1. The presentation of the decorated function in "hover" text.
2. The fact that a decorated method is treated as a static method rather than an instance method in this case.

The first issue is not a core type checker issue. It's a question of how the language server should display the decorated type when you hover over the symbol. This is subjective, but I think the current behavior is defensible. Your decorator has changed the type from a simple callable into an object that is wrapping the callable. It makes sense that the hover text would show the type of that wrapper object rather than the thing it wrapped. 

If you call this object (that is, you start typing `f(` in your editor), the language server will present you with the signature of the call along with its docstring.

<img width="327" height="111" alt="Image" src="https://github.com/user-attachments/assets/ef89e127-c389-4de8-ba85-c1762446c2c4" />

The second issue is related to an [active discussion](https://discuss.python.org/t/when-should-we-assume-callable-types-are-method-descriptors/92938/20) on the typing forum. It's an area where the official Python typing spec is ambiguous, so type checkers are left without guidance on how to handle this case. 

Pyright's current behavior is "by design". I understand that it doesn't meet your needs in the above use case, but it works well for other use cases, and it would break those use cases if I were to change it. It's not clear that there a single "right answer" that will make everyone happy.

I don't plan to change pyright's behavior here unless and until the typing spec is updated to provide clear guidance. Changing pyright's behavior prior to that would cause churn for pyright users who are relying on the current behavior, and I might need to change it back depending on the eventual resolution of the discussion thread. If the typing spec is updated and it specifies behaviors that differ from pyright's current behavior, I will implement the change at that time. 

If you have opinions about how this should be worded in the spec, feel free to contribute to the discussion in the typing forum.

**Comment by nicolasdespres** [comment]
Thank you for your quick and complete answer.

The work around I have chosen at the moment is to type the decorated function as a regular Callable (no extra attribute) and wrap access to the extra attribute into a function that dynamically checks its presence.
